src/cheri_ptw.sail

```diff
@@ -24,22 +24,58 @@ function ptw_error_to_str(e) =
 
 overload to_str = {ptw_error_to_str}
 
+/*
+ * I apologize for the confusion you have experienced if you have skipped over
+ * this comment and read the body of this function first.  Hopefully these two
+ * points resolve most of it:
+ *
+ * 1) While it may look like we're signaling that we must *always* raise
+ * faults, translationException below can see whether the instruction
+ * triggering the PTW is claiming to store an asserted tag, and will ignore our
+ * result if that isn't so.
+ *
+ * 2) We analyse only the store-side component of ext_ptw here.  The PTW logic
+ * cannot see into the data path, and so load-side responses, including traps,
+ * must be handled by the instruction that initiated the translation.
+ */
 function ext_get_ptw_error(ext_ptw : ext_ptw) -> PTW_Error =
-  match (ext_ptw) {
-    PTW_CAP_OK        => PTW_No_Permission(),
-    PTW_LOAD_CAP_ERR  => PTW_Ext_Error(AT_CAP_ERR),
-    PTW_STORE_CAP_ERR => PTW_Ext_Error(AT_CAP_ERR)
+  match (ext_ptw.ptw_sc_mod) {
+
+    /*
+     * This is confusing: PTW_No_Permission() is never used as such, but just
+     * dodges the pattern matches in translationException, below.
+     */
+    PTW_SCM_OK   => PTW_No_Permission(),
```

Comment:
Well, yes and no. `PTW_No_Permission` is just the non-extension "page fault" error. Since `ext_get_ptw_error` has been called, we know by this point that translation has failed at a PTE, so if it's not a CHERI page fault then we know it must just be a normal one. It's not so much dodging pattern matches as being one of the many conditions that lead to a standard page fault (`PTW_Access` is the only case where a failing PTW can give a non-page fault trap); really, `translationException` is saying "{Fetch,Load,Store/AMO} page fault except for in a few special cases".